orttraining/orttraining/python/training/_ortmodule_io.py

```diff
@@ -5,6 +5,52 @@
 import torch
 import warnings
 
+
+class _InputInfo(object):
+    def __init__(self, names, shape, require_grad_names=None, dynamic_axes=None, schema=None):
+        self.names = names
+        self.shape = shape
+        self.require_grad_names = require_grad_names if require_grad_names else []
+        self.dynamic_axes = dynamic_axes if dynamic_axes else {}
+        self.schema = schema if schema else []
+
+
+def _convert_input_to_list(param_names, user_input_names, buffer_names, *inputs, **kwargs):
+    '''Creates forward `*inputs` list from user input and PyTorch initializers
+
+    TODO: How IO binding model inputs and outputs affects initializer copies?
+
+    ONNX Runtime forward requires an ordered list of:
+        * User input: computed from forward InferenceSession
+        * Initializers: computed from original PyTorch model parameters
+    '''
+    # User inputs
+    non_none_inputs = [inp for inp in inputs if inp is not None]
+    named_buffers_iter = iter(buffer_names)
+    result = []
+    for input_idx, name in enumerate(user_input_names):
+        inp = None
+        if input_idx < len(non_none_inputs):
+            inp = non_none_inputs[input_idx]
+        elif name in kwargs and kwargs[name] is not None:
+            inp = kwargs[name]
+        elif input_idx >= len(non_none_inputs):
+            # Registered buffers are translated to user_input+initializer in ONNX
+            # TODO: Check what happens when the number of inputs change form one call to the next
+            buffer_name, inp = next(named_buffers_iter)
+            assert buffer_name == name, f'Input name {name} expected, but {buffer_name} found!'
+
+        if inp is not None:
+            result.append(inp)
+        else:
+            # TODO: Re-export ONNX if any input from _onnx_graphs_info.user_input_names is None.
```

Comment:
I think we probably can remove this TODO and exception raising (change it to an assert) also?
Basically, when we get to this stage, there should be no None input, right?

Reply:
Not entirely sure. I will remove the todo but keep the exception until we narrow all cases down